From dde83ee19ee8f5d00ee7db5389b9d7c8bc77a614 Mon Sep 17 00:00:00 2001
From: Alexandru Croitor <alexandru.croitor@qt.io>
Date: Fri, 15 Aug 2025 18:51:04 +0200
Subject: [PATCH 118/553] CMake: Move SBOM deferred finalization to the extend
 function

This ensures that both _qt_internal_extend_sbom and
_qt_internal_add_sbom will trigger sbom finalization for targets.

Previously only _qt_internal_add_sbom would run the finalization,
which is counter-intuitive.

One caveat is that we need to ensure we don't run finalization for
system libraries as a result of the change. Finalization of system
libraries is handled in a different pre-existing code path.

We also shouldn't call _qt_internal_extend_sbom for
qt_internal_extend_target if there are no SBOM, otherwise we trigger
SBOM creation for targets like FooModulePrivate, which will trigger an
error later because they don't have a TYPE. This might be revisited
later if it makes sense to generate SBOM entries for such targets.

Pick-to: 6.9 6.8
Task-number: QTBUG-134894
Change-Id: Ic14653f6baa920da46617061d8701ed9b94df093
Reviewed-by: Joerg Bornemann <joerg.bornemann@qt.io>
(cherry picked from commit 2088a1fff3926f723c03b96ebb9961a80ccf1f28)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 cmake/QtPublicSbomFileHelpers.cmake |  3 +-
 cmake/QtPublicSbomHelpers.cmake     | 55 ++++++++++++++++++++---------
 cmake/QtTargetHelpers.cmake         |  7 +++-
 3 files changed, 47 insertions(+), 18 deletions(-)

diff --git a/cmake/QtPublicSbomFileHelpers.cmake b/cmake/QtPublicSbomFileHelpers.cmake
index 7863a473159..52bf76148f6 100644
--- a/cmake/QtPublicSbomFileHelpers.cmake
+++ b/cmake/QtPublicSbomFileHelpers.cmake
@@ -61,7 +61,8 @@ function(_qt_internal_sbom_handle_target_binary_files target)
     )
 
     if(NOT arg_TYPE IN_LIST supported_types)
-        message(FATAL_ERROR "Unsupported target TYPE for SBOM creation: ${arg_TYPE}")
+        message(FATAL_ERROR
+            "Unsupported target TYPE '${arg_TYPE}' for target '${target}' during SBOM creation.")
     endif()
 
     set(types_without_binary_files
diff --git a/cmake/QtPublicSbomHelpers.cmake b/cmake/QtPublicSbomHelpers.cmake
index a49601afcb7..144d462acea 100644
--- a/cmake/QtPublicSbomHelpers.cmake
+++ b/cmake/QtPublicSbomHelpers.cmake
@@ -1206,25 +1206,17 @@ function(_qt_internal_add_sbom target)
 
     set(forward_args ${ARGN})
 
-    # Remove the IMMEDIATE_FINALIZATION from the forwarded args.
-    list(REMOVE_ITEM forward_args IMMEDIATE_FINALIZATION)
-
     # If a target doesn't exist we create it.
     if(NOT TARGET "${target}")
         _qt_internal_create_sbom_target("${target}" ${forward_args})
     endif()
 
-    # Save the passed options.
-    _qt_internal_extend_sbom("${target}" ${forward_args})
-
-    # Defer finalization. In case it was already deferred, it will be a no-op.
-    # Some targets need immediate finalization, like the PlatformInternal ones, because otherwise
-    # they would be finalized after the sbom was already generated.
-    set(immediate_finalization "")
     if(arg_IMMEDIATE_FINALIZATION)
-        set(immediate_finalization IMMEDIATE_FINALIZATION)
+        list(APPEND forward_args IMMEDIATE_FINALIZATION)
     endif()
-    _qt_internal_defer_sbom_finalization("${target}" ${immediate_finalization})
+
+    # Save the passed options.
+    _qt_internal_extend_sbom("${target}" ${forward_args})
 endfunction()
 
 # Helper to add custom sbom information for some kind of dependency that is not backed by an
@@ -1274,7 +1266,10 @@ function(_qt_internal_extend_sbom target)
             "a target first, or call the function on any other exsiting target.")
     endif()
 
-    set(opt_args "")
+    set(opt_args
+        NO_FINALIZATION
+        IMMEDIATE_FINALIZATION
+    )
     set(single_args
         TYPE
         FRIENDLY_PACKAGE_NAME
@@ -1283,6 +1278,14 @@ function(_qt_internal_extend_sbom target)
     cmake_parse_arguments(PARSE_ARGV 1 arg "${opt_args}" "${single_args}" "${multi_args}")
     # No validation on purpose, the other options will be validated later.
 
+    set(forward_args ${ARGN})
+
+    # Remove NO_FINALIZATION, IMMEDIATE_FINALIZATION from the forwarded args.
+    list(REMOVE_ITEM forward_args
+        NO_FINALIZATION
+        IMMEDIATE_FINALIZATION
+    )
+
     # Make sure a spdx id is recorded for the target right now, so it is "known" when handling
     # relationships for other targets, even if the target was not yet finalized.
     if(arg_TYPE)
@@ -1297,8 +1300,24 @@ function(_qt_internal_extend_sbom target)
             ${package_name_option}
         )
     endif()
-
-    set_property(TARGET ${target} APPEND PROPERTY _qt_finalize_sbom_args "${ARGN}")
+    get_target_property(is_system_library "${target}" _qt_internal_sbom_is_system_library)
+
+    set_property(TARGET ${target} APPEND PROPERTY _qt_finalize_sbom_args "${forward_args}")
+
+    # If requested via NO_FINALIZATION or the target is a system library, don't run finalization.
+    # This is necessary for system libraries because those are handled in special code path just
+    # before finishing project sbom generation, and finalizing them would cause issues because they
+    # don't actually have a TYPE until a later point in time.
+    if(NOT arg_NO_FINALIZATION AND NOT is_system_library)
+        # Defer finalization. In case it was already deferred, it will be a no-op.
+        # Some targets need immediate finalization, like the PlatformInternal ones,
+        # because otherwise they would be finalized after the sbom was already generated.
+        set(immediate_finalization "")
+        if(arg_IMMEDIATE_FINALIZATION)
+            set(immediate_finalization IMMEDIATE_FINALIZATION)
+        endif()
+        _qt_internal_defer_sbom_finalization("${target}" ${immediate_finalization})
+    endif()
 endfunction()
 
 # Helper to add additional sbom information to targets created by qt_find_package.
@@ -1327,7 +1346,11 @@ function(_qt_find_package_extend_sbom)
     _qt_internal_validate_all_args_are_parsed(arg)
 
     # Make sure not to forward TARGETS.
-    set(sbom_args "")
+    # Also don't enable finalization, because system libraries are handled specially in a different
+    # code path.
+    set(sbom_args
+        NO_FINALIZATION
+    )
     _qt_internal_forward_function_args(
         FORWARD_APPEND
         FORWARD_PREFIX arg
diff --git a/cmake/QtTargetHelpers.cmake b/cmake/QtTargetHelpers.cmake
index 485091e780e..2c8d79b89da 100644
--- a/cmake/QtTargetHelpers.cmake
+++ b/cmake/QtTargetHelpers.cmake
@@ -253,7 +253,12 @@ function(qt_internal_extend_target target)
                 FORWARD_MULTI
                     ${__qt_internal_sbom_multi_args}
             )
-            _qt_internal_extend_sbom(${target} ${sbom_args})
+            # Don't call extend_sbom unless we actually have any SBOM args, to prevent
+            # recording of a target if it has no SBOM info. Relevant for QtFooModulePrivate targets
+            # which don't appear in the SBOM atm.
+            if(sbom_args)
+                _qt_internal_extend_sbom(${target} ${sbom_args})
+            endif()
         endif()
 
         set(target_private "${target}Private")
-- 
2.51.2

